src/Expansion.php

```diff
@@ -332,6 +330,49 @@ private function lookupAstOptional(Ast $label, Ast $context) : Ast {
         return $result;
     }
 
+    private function doExpanderCall(Ast $expanderAst, $cg) {
+        $function = $cg->this->compileCallable('\Yay\Dsl\Expanders\\', $expanderAst->{'* expander'}, self::E_BAD_EXPANDER);
+        $expander = new \ReflectionFunction($function);
+        if ($expander->getParameters()) {
+            if ($expander->getParameters()[0]->hasType() && Ast::class === $expander->getParameters()[0]->getType()->getName()) {
+                return $this->doAstExpanderCall($expander, $expanderAst, $cg);
+            }
+
+            return $this->doTokenStreamExpanderCall($expander, $expanderAst, $cg);
+        }
+
+        return $expander->invoke(null, $cg->engine);
+    }
+
+    private function doTokenStreamExpanderCall(\ReflectionFunction $expander, Ast $expanderAst, $cg): TokenStream {
+        return $expander->invoke(
+            $cg->this->mutate(TokenStream::fromSlice($expanderAst->{'* args'}->tokens()), $cg->context, $cg->engine),
+            $cg->engine
+        );
+    }
+
+    private function doAstExpanderCall(\ReflectionFunction $expander, Ast $expanderAst, $cg): Ast {
+        if ($expanderAst->{'args layer_arg'}) {
+            $ts = TokenStream::fromSlice($expanderAst->{'* args'}->tokens());
+            $arg = $this->labelExpansion()->parse($ts);
+            if (null === $arg || !$ts->index() instanceof NodeEnd) {
+                $this->fail(
+                    "Error passing `TokenStream<%s>` as argument to `%s` on line %d",
+                    $expanderAst->implode(),
+                    $expanderAst->implode(),
+                    $expanderAst->{'expander_sigil 0'}->line()
+                );
+            }
+            return $expander->invoke(
+                $this->lookupAst($arg->{'* label'}, $cg->context, self::E_UNDEFINED_EXPANSION),
+                $cg->engine
+            );
+        }
+        else {
+            return $expander->invoke($this->doExpanderCall($expanderAst->{'* args expander_call'}, $cg), $cg->engine);
+        }
```

Comment:
This non return branch should be unreachable, might want to throw an exception here...